core/services/pipeline/runner.go

```diff
@@ -181,76 +182,181 @@ func (r *runner) processIncompleteTaskRuns() {
+	for id, mtr := range all {
+		for _, pred := range all {
+			if !pred.taskRun.PipelineTaskSpec.SuccessorID.IsZero() && pred.taskRun.PipelineTaskSpec.SuccessorID.ValueOrZero() == int64(id) {
+				mtr.nPredecessors++
+			}
 		}
-		var spec Spec
-		err = txdb.Find(&spec, "id = ?", taskRun.PipelineRun.PipelineSpecID).Error
-		if err != nil {
-			return Result{Error: errors.Wrap(err, "unexpected error could not find pipeline spec by ID")}
+
+		if mtr.taskRun.PipelineTaskSpec.SuccessorID.IsZero() {
+			mtr.next = nil
+		} else {
+			mtr.next = all[int32(mtr.taskRun.PipelineTaskSpec.SuccessorID.ValueOrZero())]
 		}
 
-		// Order of precedence for task timeout:
-		// - Specific task timeout (task.TaskTimeout)
-		// - Job level task timeout (spec.MaxTaskDuration)
-		// - Node level task timeout (JobPipelineMaxTaskDuration)
-		taskTimeout, isSet := task.TaskTimeout()
-		if isSet {
-			ctx, cancel = utils.CombinedContext(r.chStop, taskTimeout)
-			defer cancel()
-		} else if spec.MaxTaskDuration != models.Interval(time.Duration(0)) {
-			ctx, cancel = utils.CombinedContext(r.chStop, time.Duration(spec.MaxTaskDuration))
-			defer cancel()
+		if mtr.nPredecessors == 0 {
+			// No predecessors so this is the first one
+			graph = append(graph, mtr)
 		}
+	}
 
-		result := task.Run(ctx, taskRun, inputs)
-		if _, is := result.Error.(FinalErrors); !is && result.Error != nil {
-			logger.Errorw("Pipeline task run errored", append(loggerFields, "error", result.Error)...)
-		} else {
-			f := append(loggerFields, "result", result.Value)
-			switch v := result.Value.(type) {
-			case []byte:
-				f = append(f, "resultString", fmt.Sprintf("%q", v))
-				f = append(f, "resultHex", fmt.Sprintf("%x", v))
+	// 3. Execute tasks using "fan in" job processing
+
+	var finalResult Result
+	var wg sync.WaitGroup
+	wg.Add(len(graph))
+
+	for _, mtr := range graph {
+		go func(m *memoryTaskRun) {
+			defer wg.Done()
+			for m != nil {
+				m.predMu.RLock()
+				nPredecessors := m.nPredecessors
+				inputs := m.inputs
+				m.predMu.RUnlock()
+				if nPredecessors > 0 {
+					// This one is still waiting another chain, abandon this
+					// goroutine and let the other handle it
+					return
+				}
+
+				var finished bool
+
+				// Avoid double execution, only one goroutine may finish the task
+				m.finishMu.Lock()
+				finished = m.finished
+				if finished {
+					m.finishMu.Unlock()
+					return
+				}
+				m.finished = true
+				m.finishMu.Unlock()
+
+				start := time.Now()
+
+				result := r.executeTaskRun(txdb, m.taskRun, inputs)
+
+				// TODO: As a further optimisation later we can consider pulling out these updates
+				if err := txdb.Exec(`UPDATE pipeline_task_runs SET output = ?, error = ?, finished_at = ? WHERE id = ?`, result.OutputsDB(), result.ErrorsDB(), time.Now(), m.taskRun.ID).Error; err != nil {
+					logger.Errorw("could not mark pipeline_task_run as finished", "err", err)
+				}
+
+				elapsed := time.Since(start)
```

Comment:
nit: we should save a `finished := time.Now()` and use that for both the update and the prom execution time